Added logistic parameters to image classifier

varunlakshmanan opened the pull request:
Attempted to my add the logistic regression's parameters to the image classifier toolkit. This is probably not correct, as I'm inexperienced.

hoytak commented on src/unity/python/turicreate/toolkits/image_classifier/image_classifier.py:
```diff
@@ -27,9 +27,23 @@
+def create(dataset, target, features=None,
+    l2_penalty=0.01, l1_penalty=0.0,
+    solver='auto', feature_rescaling=True,
+    convergence_threshold = _DEFAULT_SOLVER_OPTIONS['convergence_threshold'],
```
Currently, the solver can be several different algorithms (e.g. fista, newton, or lbfgs depending on regularization values, etc.), so let's put these in a **solver_params dictionary that can pass any number of keywords through to the solvers.

hoytak commented on src/unity/python/turicreate/toolkits/image_classifier/image_classifier.py:
```diff
@@ -496,6 +510,1759 @@ def evaluate(self, dataset, metric='auto', verbose=True, batch_size=64):
+        
+        
+# Regression model names.
+
```
Whitespace issues here.

TobyRoseman replied:
I'm not sure what this is. It looks like this is just copied from logistic_classifier.py. You don't want to copy any code from that file. You just want to pass the new parameters to [the create call for logistic regression](https://github.com/apple/turicreate/blob/86c65447504be4e4bdd3378d88957c6f1b0d4d63/src/unity/python/turicreate/toolkits/image_classifier/image_classifier.py#L164).

varunlakshmanan replied:
Yes, it's copied from logistic_classifier.py.

varunlakshmanan commented:
Just added another commit addressing what @TobyRoseman mentioned. I added the new parameters to the logistic regression create call (including the model parameter), and also included the new parameters to the initial `def create` call and set their default values. What further changes need to be made?

TobyRoseman commented on src/unity/python/turicreate/toolkits/image_classifier/image_classifier.py:
```diff
@@ -27,18 +27,32 @@
+    
     """
     Create a :class:`ImageClassifier` model.
-
```
Did you mean to delete all these blank lines? We should leave them.

varunlakshmanan replied:
I see what caused this. I copied this from the image_classifier.py file from the master branch, but when pasting code, GitHub seems to automatically remove blank lines.

TobyRoseman commented on src/unity/python/turicreate/toolkits/image_classifier/image_classifier.py:
```diff
@@ -47,74 +61,57 @@ def create(dataset, target, feature = None, model = 'resnet-50',
         Models are downloaded from the internet if not available locally. Once
         downloaded, the models are cached for future use.
-
     validation_set : SFrame, optional
```
You should add documentation for each new parameter. This text can probably largely be copied from [the docstring for logistic regression's create function](https://github.com/apple/turicreate/blob/86c65447504be4e4bdd3378d88957c6f1b0d4d63/src/unity/python/turicreate/toolkits/classifier/logistic_classifier.py#L89).

varunlakshmanan commented:
@TobyRoseman Fixed what you mentioned with another commit.

TobyRoseman commented on src/unity/python/turicreate/toolkits/image_classifier/image_classifier.py:
```diff
@@ -27,9 +27,25 @@
+'lbfgs_memory_level': 11,
+'max_iterations': 10}
+
+  def create(dataset, target, features=None, model = 'resnet-50',
```
There is indentation here. This is a syntax error in Python. The 'def' should be at the very start of the line.

TobyRoseman commented on src/unity/python/turicreate/toolkits/image_classifier/image_classifier.py:
```diff
@@ -53,6 +69,78 @@ def create(dataset, target, feature = None, model = 'resnet-50',
+        <https://apple.github.io/turicreate/docs/userguide/supervised-learning/linear-regression.html>`_)
+
+    feature_rescaling : boolean, optional
+
```
For consistency sake, let's not have a blank line between the parameter name/type and description.

varunlakshmanan replied:
@TobyRoseman Done.

TobyRoseman commented on src/unity/python/turicreate/toolkits/image_classifier/image_classifier.py:
```diff
@@ -27,9 +27,25 @@
+'lbfgs_memory_level': 11,
+'max_iterations': 10}
+
+def create(dataset, target, features=None, model = 'resnet-50',
```
The `feature` parameter has been renamed to `features` which breaks things in the body of the function. Please keep the names of the existing parameter the same.

varunlakshmanan replied:
@TobyRoseman I must not have caught that when I modelled the code off of logistic_regression.py. It's fixed now.

varunlakshmanan replied:
@TobyRoseman I must not have caught that when I modelled the code off of logistic_classifier.py. It's fixed now.

varunlakshmanan replied:
Sorry for all the repeated back and forth my errors are causing. Are there other corrections that need to be made?

TobyRoseman replied:
It looks good to me, but right now we don't know if additional changes are needed. The Travis CI build is broken for unrelated reasons. We're investigating that in #1330. Once the build is fixed, we can test these changes to find out if anything else is needed.

varunlakshmanan replied:
Good to know.

varunlakshmanan replied:
Looking at #1330, it seems like Travis CI is working again and this latest commit can be tested.

TobyRoseman replied:
@varunlakshmanan - please rebase your changes on top of master and push, so these changes can be tested.

varunlakshmanan commented on src/unity/python/turicreate/toolkits/image_classifier/image_classifier.py:
```diff
@@ -34,7 +34,8 @@
+    l2_penalty=0.01, 
+    l1_penalty=0.0,
     solver='auto', feature_rescaling=True,
     convergence_threshold = _DEFAULT_SOLVER_OPTIONS['convergence_threshold'],
```
@TobyRoseman Is this better?

varunlakshmanan commented:
@TobyRoseman I'm not sure how to rebase the changes, so I just made arbitrary formatting changes and added another commit. Does that work instead?

TobyRoseman commented:
@varunlakshmanan - Everything now looks good. I'll merge it. Thank you for your contribution.

TobyRoseman commented:
Fixes #1299

varunlakshmanan commented:
Is there anything else I need to do to get this functionality in my code (like reinstall Turi Create)? If I'm not mistaken, the `import turicreate as tc` call should fetch the latest version to be used in my script. For some reason, I'm getting an unexpected argument error when I try to create an image classifier with the new parameters merged here.

abhishekpratapa commented:
@varunlakshmanan you'll have to build turicreate to use the functionality you've coded. Even though your code is in master, you won't be able to do a `pip install turicreate` with all the updated functionality (like the feature you added) until the next release. If you build master then you'll be able to use your new functionality today.

run these commands:
1) Clone the repo if you haven't in a terminal
```git clone "https://github.com/apple/turicreate.git"```
2) Change into the turicreate directory
```cd turicreate```
3) Run configure to configure the build
```./configure```
4) Once configured change directories again into:
```cd ./debug/src/unity```
5) Then run the make command
```make -j4```
6) Change directories back into the root directory
```cd ../../..```
7) Set up your python environment with the new turicreate you just built
```source ./scripts/python_env.sh debug```
8) Now you should be able to run python and import turicreate
```python```
```import turicreate from tc```

Hope that helps!
Best 
Abhishek Pratapa

varunlakshmanan commented:
@abhishekpratapa Thanks, very helpful!
